### ISSUE TITLE
compilation fails on doc comments inside of records

### ISSUE THREAD

**Issue by zploskey** [facebook/reason#1559]
Repro:
```reason
type thing = { 
  .
  /**comment**/
  "name": string
};
```
results in the following compile-time error on Reason 1.13.7 w/ bs-platform@1.10.3 compiled (refmt gives the same error):
```
File "/home/user/project/src/doubleast_comment_in_obj.re", line 4, characters 2-8:
Error: 320: <UNKNOWN SYNTAX ERROR>
File "/home/user/project/src/doubleast_comment_in_obj.re", line 1:
Error: Error while running external preprocessor
```
On 3.0.0, refmt on this file gives
```
Error: 291: <UNKNOWN SYNTAX ERROR>
```

Note that when the non-sugared syntax is used it will compile but `refmt` fails, i.e.
```reason
type thing = Js.t {
  .
  /**comment**/
  name: string
};
```

```
$ refmt --in-place src/doubleast_comment_in_obj_desugared.re 
File "src/doubleasc_comment_in_obj_desugared.re", line 1, characters 18-19:
Error: 434: syntax error, consider adding a `;' before
```

Comment placement does not seem to matter as long as it's within the curly braces, it gives an error. Comments with a single opening `*` work fine.

**Comment by cristianoc** [comment]
+ @andreypopp

**Comment by chenglou** [comment]
Likewise for record:

```reason
type bla = {
  /** foo */
  foo: bar
}
```

**Comment by zploskey** [comment]
On 3.0.1, doc comments in a Js.t object gives:
```
Error: 3053: <UNKNOWN SYNTAX ERROR>
```
Using a doc comment in a record gives:
```
Error: 2813: <UNKNOWN SYNTAX ERROR>
```
Can someone more familiar with the codebase tell me where the error codes are coming from and where might be a good place to look at to track down the problem? I guess this would probably be in the parser but I could waste a lot of time trying understand what's going on in there. I'm currently looking at places where Reason_parser_explain.message (the function that prints `<UNKNOWN SYNTAX ERROR>`) gets called and trying to trace back what's going on from there.

**Comment by zploskey** [comment]
I spoke with @IwanKaramazow about this, and we determined that this is the result of `record_expr` in the parser not allowing attributes, since docstrings are parsed as attributes. This differs from the way that OCaml parses docstrings, where they are given their own type instead of `string` and that can be either attached or unattached, and therefor don't need to be specially handled in the parser. Is there a reason we aren't handling docstrings similarly? This is the code in question:
https://github.com/facebook/reason/blob/985f28150026fb3ce353c699a2a0cbc9a52de919/src/reason-parser/reason_parser.mly#L3112-L3124
I guess this also means that it is not currently possible to annotate a particular record label/row, but annotating the corresponding expression of course works. Presumably when adding this we'll need to disallow non-docstring attributes.

I have been playing around with how to fix this but any tips would be appreciated.

**Comment by zploskey** [comment]
One thing that currently is allowed is this, because expressions can have annotations,
```reason
type rec_doc_comment_in_expr = {foo: string /** docstring */};
```
but strangely it refmts to:
```reason
type record_with_doc_comment = {foo: string /*** docstring */};
```
where it adds an extra `*` in the comment opening.

**Comment by let-def** [comment]
Right now when a comment starting with `/**` is not in an attribute position, it is printed with an added `*` so that it is not confused with a doc comment.

**Comment by zploskey** [comment]
Fixed in #1817.
